source/simulator/assembly.cc

```diff
@@ -1419,7 +1419,8 @@ namespace aspect
                    * on the direction of the flow. Thus the flow-dependent terms below are
                    * only calculated if the edge is an inflow edge.
                    */
-                  const bool inflow = ((current_u * scratch.face_finite_element_values->normal_vector(q)) < 0.);
+                  const bool inflow = (advection_field.is_temperature()
+                                       || ((current_u * scratch.face_finite_element_values->normal_vector(q)) < 0.));
```

Comment:
Thinking about this some more, shouldn't we treat Dirichlet boundary conditions via the standard explicit jump terms like we do on faces between cells, but where one half of the jump is simply the prescribed boundary term? In other words, specifically *not* use flux terms for Dirichlet values? This would mean doing the *opposite* from what this part of the patch does.